### PR TITLE
Endpoints for API keys handling

### PR DESCRIPTION
**Pull request by piotrkosecki**
Added checking for API keys usage per user to be secured with API keys like for Conseil calls.
Added endpoints for activating/deactivating keys for user.

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/doobie/ApiKeyQueries.scala [hunk 1]
```diff
@@ -43,6 +43,10 @@ trait ApiKeyQueries extends EnvironmentMappers {
   def invalidateApiKeysQuery(userId: UserId, now: Instant): Update0 =
     sql"UPDATE api_keys SET datesuspended = $now WHERE userid = $userId".update
 
+  /** Inserts API key for user */
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/nautilus/cloud/adapters/doobie/ApiKeyQueries.scala:46]: Not sure if the doc corresponds with the code

File: src/main/scala/tech/cryptonomic/nautilus/cloud/domain/authentication/AuthorizationService.scala [hunk 2]
```diff
@@ -16,4 +17,26 @@ object AuthorizationService {
       f.map(Right(_))
     else
       AccessDenied(requiredRole, session).asLeft[T].pure[F]
+
+  def requiredRoleOrApiKey[F[_]: Applicative, T](
+      requiredRole: Role,
+      apiKey: Option[String],
+      supportedKeys: List[String]
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/nautilus/cloud/domain/authentication/AuthorizationService.scala:24]: I'd suggest using sets for membership's checks, they should be better tailored to the task.

File: src/main/scala/tech/cryptonomic/nautilus/cloud/application/ApiKeyApplication.scala [hunk 3]
```diff
@@ -45,13 +52,7 @@ class ApiKeyApplication[F[_]: Monad](conseilConfig: ConseilConfig, apiKeyService
...
   /** Returns all API keys from the DB */
   def getApiKeysForEnv(apiKey: String, environment: Environment): F[Permission[List[String]]] =
-    Either
-      .cond(
-        conseilConfig.keys.contains(apiKey),
-        apiKeyService.getAllApiKeysForEnv(environment),
-        AccessDenied("Wrong API key").pure[F]
-      )
-      .bisequence
+    requiredApiKey(apiKey, conseilConfig.keys)(apiKeyService.getAllApiKeysForEnv(environment))
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/nautilus/cloud/application/ApiKeyApplication.scala:55]: Could you help me clarify: what's in conseilConfig.keys?
Is that a global key set, a specific credential-enabled set, a specific user-set?

Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/application/ApiKeyApplication.scala:55]: I don't understand what it is that it's being verified by the requiredApiKey call, in the end.

File: src/main/scala/tech/cryptonomic/nautilus/cloud/domain/metering/stats/MeteringStatsRepository.scala [hunk 4]
```diff
@@ -14,6 +16,6 @@ trait MeteringStatsRepository[F[_]] {
   def getLastStats(list: List[UserId]): F[List[AggregatedMeteringStats]]
 
   /** Fetches stats for the given user */
-  def getStatsPerUser(userId: UserId): F[List[AggregatedMeteringStats]]
+  def getStatsPerUser(userId: UserId, from: Option[Instant]): F[List[AggregatedMeteringStats]]
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/nautilus/cloud/domain/metering/stats/MeteringStatsRepository.scala:19]: Would it be useful to put a default value of `None` here, and replicate it to any implementation?

**Review by ivanopagano — COMMENTED** [review]
Minor comments and couple of questions left.

File: src/main/scala/tech/cryptonomic/nautilus/cloud/application/ApiKeyApplication.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/application/ApiKeyApplication.scala:55]: `conseilConfig.keys` are apiKeys used to validate http calls between services. Some of endpoints can work with validation by apiKeys and session, and some only with validation by session.

File: src/main/scala/tech/cryptonomic/nautilus/cloud/domain/authentication/AuthorizationService.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/nautilus/cloud/domain/authentication/AuthorizationService.scala]: Ok, changed. Just for the sake of argument - in theory it's a a good idea to perform membership checks, but the time saved by this change won't ever surpass the time I spent writing this comment 😛

**Review by ivanopagano — APPROVED** [review]

**Review by bhaskarkishore — APPROVED** [review]

**Pull request merged**